### PR TITLE
nested container types

### LINKED ISSUE #427: Nested Containers in structured config type annotation
**Describe the bug**
Using Dict[str, List[str]] or other similar complex value types fails.

**To Reproduce**
```python
from dataclasses import dataclass
from typing import Dict, Tuple, List

from omegaconf import OmegaConf, MISSING

@dataclass
class Config:
    # If I change List[str] to the following, same error:
    #   Tuple[str, str]
    #   Tuple[str, ...]
    #   Dict[str, str]
    data: Dict[str, List[str]] = MISSING

# omegaconf.errors.ValidationError: Unsupported value type : typing.List[str]
OmegaConf.structured(Config)
```

**Expected behavior**
It should work.

Instead you get Validation error (on 2.0) and the following on master:
```
Traceback (most recent call last):
  File "test.py", line 15, in <module>
    OmegaConf.structured(Config)
  File "/Users/gibiansky/env/lib/python3.7/site-packages/omegaconf/omegaconf.py", line 145, in structured
    return OmegaConf.create(obj, parent, flags)
  File "/Users/gibiansky/env/lib/python3.7/site-packages/omegaconf/omegaconf.py", line 201, in create
    flags=flags,
  File "/Users/gibiansky/env/lib/python3.7/site-packages/omegaconf/omegaconf.py", line 254, in _create_impl
    flags=flags,
  File "/Users/gibiansky/env/lib/python3.7/site-packages/omegaconf/dictconfig.py", line 105, in __init__
    format_and_raise(node=None, key=None, value=None, cause=ex, msg=str(ex))
  File "/Users/gibiansky/env/lib/python3.7/site-packages/omegaconf/dictconfig.py", line 89, in __init__
    self._set_value(content, flags=flags)
  File "/Users/gibiansky/env/lib/python3.7/site-packages/omegaconf/dictconfig.py", line 552, in _set_value
    raise e
  File "/Users/gibiansky/env/lib/python3.7/site-packages/omegaconf/dictconfig.py", line 549, in _set_value
    self._set_value_impl(value, flags)
  File "/Users/gibiansky/env/lib/python3.7/site-packages/omegaconf/dictconfig.py", line 580, in _set_value_impl
    allow_objects=self._get_flag("allow_objects"),
  File "/Users/gibiansky/env/lib/python3.7/site-packages/omegaconf/_utils.py", line 312, in get_structured_config_data
    return get_dataclass_data(obj, allow_objects=allow_objects)
  File "/Users/gibiansky/env/lib/python3.7/site-packages/omegaconf/_utils.py", line 253, in get_dataclass_data
    parent=dummy_parent,
  File "/Users/gibiansky/env/lib/python3.7/site-packages/omegaconf/omegaconf.py", line 846, in _maybe_wrap
    ref_type=ref_type,
  File "/Users/gibiansky/env/lib/python3.7/site-packages/omegaconf/omegaconf.py", line 777, in _node_wrap
    element_type=element_type,
  File "/Users/gibiansky/env/lib/python3.7/site-packages/omegaconf/dictconfig.py", line 105, in __init__
    format_and_raise(node=None, key=None, value=None, cause=ex, msg=str(ex))
  File "/Users/gibiansky/env/lib/python3.7/site-packages/omegaconf/dictconfig.py", line 77, in __init__
    flags=flags,
  File "<string>", line 10, in __init__
  File "/Users/gibiansky/env/lib/python3.7/site-packages/omegaconf/base.py", line 42, in __post_init__
    assert self.element_type is Any or isinstance(self.element_type, type)
AssertionError
```

Python 3.7.5 on recent Mac OSX.

### PR DESCRIPTION
**Pull request by Jasha10**
This PR implements support for nested container types.

Here's a quick demo:
```python
from dataclasses import dataclass, field
from typing import Dict, List

from pytest import raises

from omegaconf import MISSING, OmegaConf, ValidationError


@dataclass
class NestedContainers:
    dict_of_dict: Dict[str, Dict[str, int]]
    list_of_list: List[List[int]] = field(default_factory=lambda: [[123]])
    dict_of_list: Dict[str, List[int]] = MISSING
    list_of_dict: List[Dict[str, int]] = MISSING


cfg = OmegaConf.structured(NestedContainers(dict_of_dict={"foo": {"bar": 123}}))

print(OmegaConf.to_yaml(cfg))
# dict_of_dict:
#   foo:
#     bar: 123
# list_of_list:
# - - 123
# dict_of_list: ???
# list_of_dict: ???

with raises(ValidationError):
    cfg.list_of_dict = [["whoops"]]  # not a list of dicts
```

The biggest change to the source is the addition of a `_deep_update_type_hint` function to the `basecontainer.py` file. The purpose of this new function is to deeply validate compatibility of an OmegaConf object with a given type hint, mutating the object's metadata if necessary. This is necessary when merging nodes that are typed as nested containers, or when assigning an OmegaConf object to a field that is typed as a nested container.

resolves #427

### CODE REVIEW

File: omegaconf/basecontainer.py [hunk 1]
```diff
@@ -246,7 +250,7 @@ def get_node_value(key: Union[DictKeyType, int]) -> Any:
 
         if isinstance(conf, DictConfig):
             if (
-                conf._metadata.object_type is not None
+                conf._metadata.object_type not in (dict, None)
```
Comment (pixelb) [inline omegaconf/basecontainer.py:253]: for my reference, this is correct as tuple.__contains__() does "a is b or a == b"

**Review by pixelb — APPROVED** [review]
Very nice work on this, especially on the tests.
What's the python 3.7 constraint on this?
It should probably be mentioned in the docs given how common 3.6 is

**Comment by Jasha10** [comment]
> Very nice work on this, especially on the tests. What's the python 3.7 constraint on this? It should probably be mentioned in the docs given how common 3.6 is

Thanks Pádraig!

I've just pushed a commit documenting the Python3.6 limitation.
The issue is with pickling subscripted `typing`. Here are some examples:
```python
$ python3.6 -q
>>> from pickle import dumps
>>> from typing import *
>>> dumps(Dict[str, int])
...
_pickle.PicklingError: Can't pickle typing.Dict[str, int]: it's not the same object as typing.Dict
>>> dumps(DictConfig({}, element_type=Dict[str, int]))
...
_pickle.PicklingError: Can't pickle typing.Dict[str, int]: it's not the same object as typing.Dict
```
I believe this was the original motivation for OmegaConf's storing `optional` and `ref_type` metadata separately: the type hint `Optional[int]` cannot be pickled (by py3.6), but the ref_type `int` can be pickled together with a boolean flag indicating optionality.

Related is the [`BaseContainer.__getstate__`](https://github.com/omry/omegaconf/blob/f1f4a5279a613b705596aaf32210db67b732e324/omegaconf/basecontainer.py#L96-L100) function, which sidesteps some of the above errors by discarding information from subscripted dicts and lists appearing in `ref_type`.

**Review by jieru-hu — APPROVED** [review]
looks great! thanks!

File: omegaconf/dictconfig.py [hunk 2]
```diff
@@ -677,18 +680,20 @@ def _set_value_impl(
...
+                self._metadata.object_type = value._metadata.object_type
 
             elif isinstance(value, dict):
                 with flag_override(self, ["struct", "readonly"], False):
                     for k, v in value.items():
                         self.__setitem__(k, v)
+                self._metadata.object_type = dict
 
             else:  # pragma: no cover
+                # msg = f"Unsupported value type: {value} ({type(value).__name__})"
```
[the remark this answers is not in the document]
Comment (Jasha10) [reply omegaconf/dictconfig.py:696]: Nope! Good catch.

**Pull request merged**